File: examples/weather-api/typescript/bridges/weather/host.ts
// bridges/weather/host.ts — TypeScript bridge host for the weather example.
// `rusm build` generates the Rust delegation shim, the TS runner, and the host binary. The
// runner is a long-lived resident actor; logic here is synchronous — no I/O.
//
// WIT `enum` values on the wire are the variant names as the generated bindings render them —
// see the ambient `bridges.d.ts` this build writes (`Units = "Celsius" | "Fahrenheit"`,
// `Sky = "Sunny" | "Cloudy" | "Rainy"`). A guest is type-checked against those; a bridge host
// must return/accept the same spellings.

export function lookup(city: string): string {
  return `sunny in ${city}`;
}

// `query` is a WIT `record` (a plain JS object); the return matches the WIT `report` record,
// with the `sky` enum as one of the generated `Sky` values.
export function detailed(query: {
  city: string;
  units: "Celsius" | "Fahrenheit";
}): { city: string; sky: "Sunny" | "Cloudy" | "Rainy"; temp: number } {
  const temp = query.units === "Fahrenheit" ? 70 : 21;
  return { city: query.city, sky: "Sunny", temp };
}

File: examples/weather-api/go/bridges/weather/host.go
// bridges/weather/host.go — Go bridge host for the weather example.
// rusm build generates _runner.go and go.mod (if absent); TinyGo compiles the whole
// bridges/weather/ package to wasm/bridge-weather.wasm. Logic here is synchronous.
package main

import (
	"encoding/json"
	"fmt"
)

// Lookup returns a forecast string for the city.
func Lookup(city string) string {
	return fmt.Sprintf("sunny in %s", city)
}

// Detailed returns a structured forecast.
// WIT record params arrive as json.RawMessage in the generated dispatcher — unmarshal into your
// own struct for full control over field naming and validation. WIT `enum` values on the wire
// are the variant names as the bindings render them (`Units` = "Celsius"/"Fahrenheit", `Sky` =
// "Sunny"/"Cloudy"/"Rainy") — a bridge host must read/write those exact spellings.
func Detailed(raw json.RawMessage) json.RawMessage {
	var q struct {
		City  string `json:"city"`
		Units string `json:"units"`
	}
	if err := json.Unmarshal(raw, &q); err != nil {
		q.City = "unknown"
		q.Units = "Celsius"
	}
	temp := int32(21)
	if q.Units == "Fahrenheit" {
		temp = 70
	}
	out, _ := json.Marshal(map[string]any{
		"city": q.City,
		"sky":  "Sunny",
		"temp": temp,
	})
	return out
}
